File: Exercise_3.cpp
// Time Complexity                              :  O(N * M), where N is the row size and M is the column size of the matrix
// Space Complexity                             :  O(1)
// Did this code successfully run on Leetcode   :  Yes
// Any problem you faced while coding this      :  No

#include <bits/stdc++.h> 
using namespace std;  


class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        
        int n = matrix.size();
        int m = matrix[0].size();
        
        int top = 0;
        int down = n - 1;
        int left = 0;
        int right = m - 1;
        
        vector<int> ans;
        int index = 0;
        while(left <= right and top <= down) {
            
            // first travserse left-right and capture row elements in ans array.
            for(int j=left;j<=right;j++) {
                ans.push_back(matrix[top][j]);
            }
            // move to the next row since this row has been captured.
            top++;

            
            // change direction to top-down to capture the column elements in the ans array.
            for(int i=top;i<=down;i++) {
                ans.push_back(matrix[i][right]);
            }
            // move to the next column from right column since this column is captured.
            right--;
            
            // again change direction to right-left and capture all the elements in the ans array.
            if(top <= down) {
                for(int j=right;j>=left;j--){
                    ans.push_back(matrix[down][j]);
                }
                // move to a row above since it has been captured.
                down--;
            }
            
            // again change direction to down-top and capture all the elements in the ans array.
            if(left <= right) {
                for(int i=down;i>=top;i--){
                    ans.push_back(matrix[i][left]);
                }
                // move to the next column right column to the current one since we have captured this column.
                left++;
            }
        
        }    
        
        return ans;
        
    }
        
};

File: Exercise_3.js
// Time Complexity : O(m * n) where m*n is matrix
// Space Complexity : O(m * n)
// Approach : transversing on matrix from left to right than top to bottom than right to left than bottom to top 

/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function (matrix) {
    const resultArr = [];

    if (!matrix || matrix.length === 0 || matrix[0].length === 0) {
        return resultArr;
    }

    let top = 0;
    let right = matrix[0].length - 1;
    let bottom = matrix.length - 1;
    let left = 0;

    while (left <= right && top <= bottom) {
        // Traverse right
        for (let j = left; j <= right; j++) {
            resultArr.push(matrix[top][j]);
        }
        top++;

        // Traverse down
        for (let i = top; i <= bottom; i++) {
            resultArr.push(matrix[i][right]);
        }
        right--;

        // Traverse left
        if (top <= bottom) {
            for (let j = right; j >= left; j--) {
                resultArr.push(matrix[bottom][j]);
            }
            bottom--;
        }

        // Traverse up
        if (left <= right) {
            for (let i = bottom; i >= top; i--) {
                resultArr.push(matrix[i][left]);
            }
            left++;
        }
    }
    return resultArr;
};
